joblib/numpy_pickle.py

```diff
@@ -329,44 +338,49 @@ def load_build(self):
         dispatch[pickle.BUILD] = load_build
 
 
-class ZipNumpyUnpickler(NumpyUnpickler):
-    """A subclass of our Unpickler to unpickle on the fly from
-    compressed storage."""
-
-    def __init__(self, filename, file_handle):
-        NumpyUnpickler.__init__(self, filename,
-                                file_handle,
-                                mmap_mode=None)
-
-    def _open_pickle(self, file_handle):
-        return BytesIO(read_zfile(file_handle))
+def _dump(value, f, protocol):
+    """Helper function that wraps the call to joblib internal pickler."""
+    pickler = NumpyPickler(f, protocol=protocol)
+    pickler.dump(value)
 
 
 ###############################################################################
 # Utility functions
 
-def dump(value, filename, compress=0, cache_size=100, protocol=None):
-    """Fast persistence of an arbitrary Python object into one or multiple
-    files, with dedicated storage for numpy arrays.
+def dump(value, filename, compress=0, protocol=None, compress_method='zlib',
+         cache_size=None):
+    """Persist an arbitrary Python object.
+
+    Fast persistence of an arbitrary Python object into one file with
+    dedicated storage for numpy arrays.
 
     Parameters
     -----------
     value: any Python object
         The object to store to disk
-    filename: string
+    filename: str
         The name of the file in which it is to be stored
-    compress: integer for 0 to 9, optional
-        Optional compression level for the data. 0 is no compression.
+    compress: int from 0 to 9 or bool or 2-uple, optional
+        Optional compression level for the data. 0 or False is no compression.
         Higher means more compression, but also slower read and
         write times. Using a value of 3 is often a good compromise.
         See the notes for more details.
-    cache_size: positive number, optional
-        Fixes the order of magnitude (in megabytes) of the cache used
-        for in-memory compression. Note that this is just an order of
-        magnitude estimate and that for big arrays, the code will go
-        over this value at dump and at load time.
+        If compress is True, the compression level used is 3.
+        If compress is a 2-uple, the first element must correspond to a string
+        between supported compressors (e.g 'zlib', 'gzip', 'bz2', 'lzma'
+        'xz'), the second element must be an integer from 0 to 9, corresponding
+        to the compression level.
+        compress_method option is ignored when using a 2-uple in compress.
     protocol: positive int
         Pickle protocol, see pickle.dump documentation for more details.
+    compress_method: {'zlib', 'gzip', 'bz2', 'lzma', 'xz'}, optional
```

Comment:
It seems like `compress_method` and `compress` as a 2-tuple gives two ways of achieving the same thing. I think we should agree on what the best interface is and use only this one.

Possible candidates:
* compress as a 2-tuple, e.g. ('bz2', 6) as implemented currently
* compress as a string, e.g 'bz26', doesn't look that great for 'bz2' actually, maybe we could add an underscore, i.e. 'bz2_6'. @GaelVaroquaux was proposing 'bz2[6]' in https://github.com/joblib/joblib/issues/273#issuecomment-157533505.
* two arguments, `compress_method` and `compress`. The problem is then we have to decide what is the default for each compression method, i.e. what happens to compress for compress=True.
* something better I haven't thought of

@ogrisel @GaelVaroquaux any thoughts on this?